Feature init django app

kbeker opened the pull request.

rwrzesien commented on coverage-config:
```diff
@@ -0,0 +1,12 @@
+[run]
+omit=
+	manage.py
```
Please change to 4 spaces.

kbeker replied:
Done

kbeker replied:
Done

rwrzesien commented on project_liberation/project_liberation/settings/__init__.py:
```diff
@@ -0,0 +1,14 @@
+try:
```
Is this something you have added or now Django auto generates this?

cameel replied:
This looks like my usual code for importing `local_settings.py`. It's not generated by Django.

@kbeker It would be more helpful if you had the files generated by Django and the custom changes made on top of them in separate commits. Then every change could have a description explaining why you're doing it. As it is now, it's very hard to see what is standard stuff and what is not. For example I just skimmed over the settings file and I have no idea if you added something there or not - checking that manually would be just too time consuming. Good split into commits is essential to make this review easy.

kbeker replied:
Sure, I will change it. My bad :)

rwrzesien commented on project_liberation/project_liberation/settings/development.py:
```diff
@@ -0,0 +1,12 @@
+from .base import *
+
+ENVIRONMENT = "development"
```
What do we need this setting for?

cameel replied:
This seems to have been copied from our other projects. In the projects I initiate I always define this setting so that you can check what environment you're running in if you have to.

The concept came from Ruby on Rails which had a similar setting but to be honest I never actually had a real need to use it. I always prefer to add a new setting specific to the situation.

rwrzesien replied:
I am just afraid that someone might be tempted to use this in application logic :)

cameel replied:
Then just remove it. I'm only explaining how it got here.

kbeker replied:
Okay so I'm deleting this.

rwrzesien commented on project_liberation/requirements.lock:
```diff
@@ -0,0 +1,36 @@
+appdirs==1.4.3
```
Please move both requirements to upper dir.

kbeker replied:
You already changed it to pipenv ;)

rwrzesien commented:
Looks good, just some details to clarify. I guess there is no rush with this one?

cameel commented on project_liberation/requirements.txt:
```diff
@@ -0,0 +1,9 @@
+flake8
+mypy
+psycopg2
+pylint
```
Why no separate test/development requirements?

kbeker replied:
Already changed by @rwrzesien

cameel commented on project_liberation/requirements.txt:
```diff
@@ -0,0 +1,9 @@
+black
+Django
```
Is the project really prepared to run with any version of Django? No constraints at all?

kbeker replied:
For this moment I don't want to add any constraints. If any future update will make some problems then I will add constraint to version.

cameel commented on project_liberation/requirements.txt:
```diff
@@ -0,0 +1,9 @@
+black
```
Why not [pipenv](https://docs.pipenv.org)?

rwrzesien replied:
Have you had a chance to use it already?

cameel replied:
I used it for a brief time at the beginning of one of our projects but back then it had too many issues (e.g. it was dumping system-specific debug info into the lock file). But these issues have been resolved for a long time now and I've been trying to encourage people to try it several times. Everyone complains about the requirements file and people do the same mistakes over and over again but for some weird reason no one wants to try using a better tool when given a chance to do so in a new project.

rwrzesien replied:
OK, then I would be for using it too. Just wanted to know if you have used it already so we don't run into some problems later.

rwrzesien replied:
I have replaced requirements with pipfiles to try it, it has some problems with installing `black` but it is known problem https://github.com/microsoft/vscode-python/issues/5171. After that, it is working on fresh virtualenv, so I have added it to this branch in separate commit.

cameel replied:
I think this thread is more relevant: https://github.com/python/black/issues/517. The VSCode solution seems to be to change the way you run pipenv (i.e. to run it twice and add `--pre` flag when running it just for black if I understand correctly?). Pinning to a specific version as suggested in the thread above seems like a better workaround to me (i.e. something like `black = "==18.9b0"`).

Anyway, this is not specific to black - any dependency. without a stable release will have this issue. But it's weird that they don't have one at this stage.

cameel replied:
I see that you actually used `allow_prereleases`. I don't think it's a good idea in general. At most as a workaround but then it should have a comment pointing out that it's only temporary. Anyway, I'd personally just pin the version.

rwrzesien replied:
I get some weird results with this tool. Once suddenly, today it started working without `allow_prereleases` on fresh venv. Then, repeating the process, I get back to the original error.

At some point when I have set specific version for `black` I get to the error 
```
pipenv.patched.notpip._internal.exceptions.UnsupportedPythonVersion: black requires Python '>=3.6' but the running Python is 2.7.12
``` 

when I have set it to `==19.3b0` even when I am on virtualenv with Python 3.6. Setting back to `black = "*"` or `==18.9b0` allowed me to install. 

For now it works on fresh venv but I am a little bit sceptical if we won't run into some problems with it.

cameel replied:
Maybe it's actually a problem with black? Or at least with the newest release?

Anyway, I don't really know how reliable pipenv is but the best way to find out is to start using it. A small project like this is the perfect opportunity to do so. If it turns out it has serious drawbacks, we won't be using it for anything else. But if it's good, we'll finally be able to stop doing the tedious and error-prone dependency management manually so it's worth trying.

By the way, I was under impression that pipenv was becoming a standard tool in the Python community but seems like lots of people don't like it for various reasons. The most common arguments against it are related to the fact that it does not replace pip and does not work well for libraries - which is not a problem for us since we have an application here. You could try one of the alternatives:
- [poetry](https://poetry.eustace.io/)
- [hatch](https://github.com/ofek/hatch)
- [flit](https://github.com/takluyver/flit)

rwrzesien replied:
I am not saying I don't like it, just that I was very excited at the beginning, and now I am little less. But lets keep it. Maybe its problem with black itself as you have said.

cameel commented on coverage-config:
```diff
@@ -0,0 +1,12 @@
+    	*/wsgi.py
+    	*/migrations/*
+    	*/mocks/*
+    	*/tests*
```
You forgot `/` here.

kbeker replied:
Done

kbeker replied:
Done

cameel commented on lint.sh:
```diff
@@ -0,0 +1,16 @@
+printf "[PYLINT: project-liberation]\n"
+# Find all subdirectories of our python apps and use xargs to pass them as arguments to pylint
+
+./find-files-to-refactor.sh | xargs pylint --rcfile=pylintrc
```
What do you mean by "files to refactor"? Are you really going to be refactoring them? Why not run pylint on all relevant files?

kbeker replied:
This name better refer to what black and isort do with this files. I change it to more proper name.

kbeker replied:
Name changed. I think this name is more relevant

kbeker commented on Pipfile:
```diff
@@ -0,0 +1,20 @@
+psycopg2 = "*"
+
+[requires]
+python_version = "3.6"
```
I would change it to `3.7` :) @rwrzesien are you using `Ubuntu`?

rwrzesien replied:
Yes, but if you want we can update, I will just have to add another Python version, no problem with that (or at least none that I expect). Or leave it like this for now if you want.

kbeker commented:
@rwrzesien and @cameel : I had to do force push because I changed all commits in this branch. I fixed all your comments and if you can please review it again. :)

kbeker commented:
Okay, so I'm merging it into master. Closing
